jax/interpreters/xla.py

```diff
@@ -601,27 +601,50 @@ def _xla_callable_device(nreps, backend, device, arg_devices):
     else:
       assert False  # Unreachable given the error check in _xla_callable
 
-def _xla_callable_args(c, avals, tuple_args, replicated=None):
+# Used within _xla_callable_args and _xla_param to distinguish between None (no
+# sharding annotation set) and replicated.
+_replicated_param = object()
```

Comment:
I think the situation is pretty similar to replication, but we've solved it in different ways (and neither of them feels especially clean): the underlying XLA API takes an empty std::vector to signal "no annotation". I don't have a great solution, and the current code LGTM, but I feel like there's got to be a more coherent way to deal with both replication and partition annotations in this set of functions...

Reply:
Yeahh this kinda sucks. I added this because I need to distinguish between three options for what to do with a single parameter in the non-tuple case: partitioned (tuple of ints), replicated (this thing), no annotation (None). I don't need it elsewhere because everywhere else deals with sequences of values, so no annotation is None and everything else is a sequence. I think replication already has True, False, None as options so it's doesn't need this stupid extra option.

If I push the trinary logic into xla_bridge, this might be a little cleaner. But I still will probably need the extra replicated option, and then it'll be everywhere instead of in this one internal place... so I'm not sure. I'm gonna try it in a follow-up change along with the context manager change George suggested.